src/aspire/utils/rotation.py

```diff
@@ -0,0 +1,280 @@
+"""
+Define a Rotation Class for customized rotation operations used by ASPIRE.
+"""
+
+import copy
+
+import numpy as np
+from numpy.linalg import norm
+from scipy.linalg import svd
+from scipy.spatial.transform import Rotation as sp_rot
+
+from aspire.utils import ensure
+from aspire.utils.random import Random
+
+
+class Rotation:
+    def __init__(self, num_rots, seed=None, seq="ZYZ", angles=None, dtype=np.float32):
+        """
+        Initialize a Rotation object
+
+        :param num_rots: Total number of rotation sets
+        :param seed: Seed for initializing random rotations.
+            If None, the random generator is not re-seeded.
+        :param seq:  Sequence of order applying Euler angles
+        :param angles: Euler angles (in degrees) to generate rotation matrices.
+            If None, uniformly distributed angles will be generated.
+        :param dtype: data type for angles and rotation matrices
+        """
+        self.rot_seq = seq
+        self.dtype = np.dtype(dtype)
+        if angles:
+            ensure(
+                num_rots == angles.shape[0],
+                "Number of rotation matrices should equal to "
+                "number of sets of Euler angles.",
+            )
+            self.angles = angles.astype(self.dtype)
+        else:
+            self.angles = self._uniform_random_angles(num_rots, seed=seed)
+        self.num_rots = num_rots
+
+        self.data = self.rot_matrices
+        self.shape = (num_rots, 3, 3)
+
+    def __str__(self):
+        """
+        String representation.
+        """
+        return (
+            f"Rotation object with matrices[{self.num_rots}, 3, 3] of {self.dtype} type"
```

Comment:
No need to mention matrices. Just "Rotation stack consisting of {self.num_rots} elements of {self.dtype} type".

Reply:
Done.